vulcain.go

```diff
@@ -0,0 +1,283 @@
+// Package vulcain helps implementing the Vulcain protocol (https://vulcain.rocks).
+// It provides helper functions to parse HTTP requests containing "preload" and "fields" directives,
+// to extract and push using HTTP/2 Server Push the relations of a JSON document matched by the "preload" directive,
+// and to modify the JSON document according to both directives.
+//
+// The package can be used in conjunction with Go's httputil.ReverseProxy as well as in any HTTP handler.
+package vulcain
+
+import (
+	"errors"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strconv"
+
+	"github.com/dunglas/httpsfv"
+	"github.com/getkin/kin-openapi/openapi3filter"
+	"go.uber.org/zap"
+)
+
+var (
+	jsonRe   = regexp.MustCompile(`(?i)\bjson\b`)
+	preferRe = regexp.MustCompile(`\s*selector="?json-pointer"?`)
+)
+
+// Options allows to set the configuration
+type Options struct {
+	// OpenAPIFile contains the path to an OpenAPI definition (in YAML or JSON) documenting the relation between resources
+	// This is useful only for non-hypermedia APIs
+	OpenAPIFile string
+
+	// MaxPushes is the maximum number of resources to push
+	// Set MaxPushes to -1 to disable the limit
+	MaxPushes int
+
+	// Logger is the logger to use
+	Logger *zap.Logger
+}
+
+// Vulcain parses and allow to modify the HTTP response according to the "preload" and "fields" directives extracted from the request
+// Use New to create a Vulcain instance
+type Vulcain struct {
+	pushers *pushers
+	openAPI *openAPI
+	logger  *zap.Logger
+}
+
+// New creates a Vulcain instance
+func New(options Options) *Vulcain {
```

Comment:
Looks like this constructor could utilize [functional options](https://sagikazarmark.hu/blog/functional-options-on-steroids/) since most of the options are actually optional.

Reply:
Nice! I'll do that.

Reply:
Done in de33341a2ca258cee1a2be11c98776824e1602a0. Very nice pattern!

Reply:
Yeah, I like it very much!